In [1]:
# imports
from PIL import Image, ImageColor
from sklearn.metrics import confusion_matrix, classification_report
import os
import numpy as np
path = './HW1_Q6_Dataset/Q6_Dataset/Images/'

In [2]:
# image paths
samples = os.listdir(path)

In [3]:
# classifier takes the mean of the red and blue for each pixel and the compares them
# if red mean is higher then predicted class is red and if blue mean if higher the class is blue
def classifier(s):
    img = Image.open(os.path.join(path, s))
    img = np.array(img)    
    mean_red = np.mean(img[:, :, 0])
    mean_blue = np.mean(img[:, :, 2])
    if mean_red > mean_blue:
        return 'm'
    else:
        return 'c'

In [4]:
y_true = []
y_pred = []

for i, s in enumerate(samples):
    y_true.append(s[0])
    y_pred.append(classifier(s))

### Obtaining confusion matrix

In [5]:
confusion_mtx = confusion_matrix(y_true, y_pred)
confusion_mtx

array([[46, 18],
       [ 1, 57]])

As we can see above:
- 46+18 samples are actually chelsea --> 48 of them are predicted correctly and 18 of them are predicted wrongly.

- 57 + 1 samples are actually manchester --> 57 of them are predicted correctly and 1 of them are predicted wrongly.

In [6]:
correct_classification = np.sum(np.diag(confusion_mtx))
correct_classification

103

### Obtaining metricsmatrix

In [7]:
print(classification_report(y_true, y_pred, target_names=['chelsea', 'manchester']))

              precision    recall  f1-score   support

     chelsea       0.98      0.72      0.83        64
  manchester       0.76      0.98      0.86        58

    accuracy                           0.84       122
   macro avg       0.87      0.85      0.84       122
weighted avg       0.87      0.84      0.84       122



As we can see above:
- Accuracy is 84% it means that each prediction is correct by 84% probability.
- Recall for chelsea is 72%. It means that among all real chelsea samples, 72% of them are predicted correctly. Similarly recall for manchester is 98% it means that among all real manchester samples, 98% of them are predicted correctly. 
- Precision for chelsea is 98% it means that among all chelsea predictions, 98% of them were actually chelsea. Similarly precision for manchester is 76% among all manchester predictions, 76% of them were actually manchester.

### Obtaining confidence matrix

In [8]:
confidence_matrix = confusion_mtx / np.sum(confusion_mtx, axis=0)
confidence_matrix

array([[0.9787234, 0.24     ],
       [0.0212766, 0.76     ]])

As it can be seen above:
- If we predict a sample to be chelsea, we have 97% confidence that the prediction is correct and 2% confidence that it is not.
- If we predict a sample to be manchester, we have 76% confidence that the prediction is correct and 24% that it is not.